xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/AudioDSP.h

```diff
@@ -0,0 +1,1288 @@
+    ///
+    /// Remove the selected id from currently used DSP processes
+    /// @param handle identification data for stream
+    /// @return AE_DSP_ERROR_NO_ERROR if the becomes found and removed
+    /// @remarks Valid implementation required.
+    ///
+    virtual AE_DSP_ERROR StreamDestroy(const ADDON_HANDLE handle) = 0;
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @brief Ask the add-on about a requested processing mode that it is
+    /// supported on the current stream. Is called about every add-on mode after
+    /// successed StreamCreate.
+    /// @param handle identification data for stream
+    /// @param type The processing mode type, see AE_DSP_MODE_TYPE for definitions
+    /// @param mode_id The mode inside add-on which must be performed on call. Id
+    /// is set from add-on by iModeNumber on AE_DSP_MODE structure during
+    /// RegisterMode callback,
+    /// @param unique_db_mode_id The Mode unique id generated from dsp database.
+    /// @return AE_DSP_ERROR_NO_ERROR if the properties were fetched successfully
+    /// or if the stream is not supported the add-on must return
+    /// AE_DSP_ERROR_IGNORE_ME.
+    /// @remarks Valid implementation required.
+    ///
+    virtual AE_DSP_ERROR StreamIsModeSupported(const ADDON_HANDLE handle, AE_DSP_MODE_TYPE type, unsigned int mode_id, int unique_db_mode_id) = 0;
+    //--------------------------------------------------------------------------
+
+    //==========================================================================
+    ///
+    /// @brief Set up Audio DSP with selected audio settings (detected on data of
+    /// first present audio packet)
+    /// @param addonSettings The add-ons audio settings.
+    /// @return AE_DSP_ERROR_NO_ERROR if the properties were fetched successfully.
+    /// @remarks Valid implementation required.
+    ///
+    virtual AE_DSP_ERROR StreamInitialize(const ADDON_HANDLE handle, const AE_DSP_SETTINGS& addonSettings) = 0;
+    //--------------------------------------------------------------------------
+
+    //@}
+
+    /// @name DSP input processing
+    ///  @remarks Only used by KODI if bSupportsInputProcess is set to true.
+    ///
+    //@{
+    //==========================================================================
+    ///
+    /// @brief DSP input processing
+    /// Can be used to have unchanged stream..
+    /// All DSP add-ons allowed to-do this.
+    /// @param handle identification data for stream
+    /// @param array_in Pointer to data memory
+    /// @param samples Amount of samples inside array_in
+    /// @return true if work was OK
+    /// @remarks Optional. Is set by AE_DSP_ADDON_CAPABILITIES and asked with
+    /// GetCapabilities
+    ///
+    virtual bool InputProcess(const ADDON_HANDLE handle, const float** array_in, unsigned int samples) { return true; }
```

Comment:
It has to be a double pointer to a an array because of the performance and how AE manages the audio stream packets. All audio packets in AE are based on [CSoundPacket](https://github.com/AchimTuran/xbmc/blob/AudioDSP/v2/core-reimplementation_V18/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAEBuffer.cpp#L34) and the [CActiveAE::AllocSoundSample](https://github.com/AchimTuran/xbmc/blob/AudioDSP/v2/core-reimplementation_V18/xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp#L3087-L3098) method. Also in future AudioDSP V2 will use this class.